File: app/src/main/java/com/popalay/cardme/business/cards/CardInteractor.java
package com.popalay.cardme.business.cards;

import com.popalay.cardme.R;
import com.popalay.cardme.business.exception.ExceptionFactory;
import com.popalay.cardme.data.models.Card;
import com.popalay.cardme.data.repositories.card.CardRepository;
import com.popalay.cardme.data.repositories.holder.HolderRepository;

import java.util.List;

import javax.inject.Inject;
import javax.inject.Singleton;

import io.card.payment.CreditCard;
import rx.Completable;
import rx.Observable;
import rx.Single;
import rx.schedulers.Schedulers;

@Singleton
public class CardInteractor {

    private final CardRepository cardRepository;
    private final HolderRepository holderRepository;

    @Inject public CardInteractor(CardRepository cardRepository, HolderRepository holderRepository) {
        this.cardRepository = cardRepository;
        this.holderRepository = holderRepository;
    }

    public Single<Card> transformCard(CreditCard creditCard) {
        final Card card = transform(creditCard);
        return cardRepository.isCardExist(card)
                .flatMap(exist -> exist ? Single.error(createCardExistError()) : Single.just(card))
                .subscribeOn(Schedulers.io());
    }

    public Completable save(Card card) {
        return cardRepository.save(card)
                .andThen(holderRepository.updateCounts(card.getHolder()))
                .subscribeOn(Schedulers.io());
    }

    public Observable<List<Card>> getCards() {
        return cardRepository.getAll().subscribeOn(Schedulers.io());
    }

    public Observable<List<Card>> getCardsByHolder(long holderId) {
        return cardRepository.getAllByHolder(holderId).subscribeOn(Schedulers.io());
    }

    public Completable updateCards(List<Card> items) {
        for (int i = 0; i < items.size(); i++) {
            items.get(i).setUsage(i);
        }
        return cardRepository.save(items)
                .subscribeOn(Schedulers.io());
    }

    public Completable removeCard(Card card) {
        return cardRepository.remove(card)
                .andThen(holderRepository.updateCounts(card.getHolder()))
                .subscribeOn(Schedulers.io());
    }

    private Card transform(CreditCard creditCard) {
        final Card card = new Card(creditCard);
        card.setGeneratedBackgroundSeed(System.nanoTime());
        return card;
    }

    private Throwable createCardExistError() {
        return ExceptionFactory.createError(ExceptionFactory.ErrorType.CARD_EXIST, R.string.error_card_exist);
    }
}

File: app/src/main/java/com/popalay/cardme/business/cards/CardInteractor.kt
package com.popalay.cardme.business.cards

import com.popalay.cardme.R
import com.popalay.cardme.business.exception.ExceptionFactory
import com.popalay.cardme.data.models.Card
import com.popalay.cardme.data.repositories.card.CardRepository
import com.popalay.cardme.data.repositories.holder.HolderRepository
import io.reactivex.Completable
import io.reactivex.Flowable
import io.reactivex.Single
import io.reactivex.schedulers.Schedulers
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class CardInteractor @Inject constructor(
        private val cardRepository: CardRepository,
        private val holderRepository: HolderRepository
) {

    fun validateCard(creditCard: Card): Single<Card> {
        val card = transform(creditCard)
        return cardRepository.cardIsNew(card)
                .flatMap { if (it) Single.just(card) else Single.error<Card>(createCardExistError()) }
                .subscribeOn(Schedulers.io())
    }

    fun save(card: Card): Completable = cardRepository.save(card)
            .andThen(holderRepository.updateCounts(card.holder))
            .subscribeOn(Schedulers.io())

    fun getCards(): Flowable<List<Card>> = cardRepository.getAll().subscribeOn(Schedulers.io())

    fun getCardsByHolder(holderId: String): Flowable<List<Card>> = cardRepository.getAllByHolder(holderId)
            .subscribeOn(Schedulers.io())

    fun updateCards(items: List<Card>): Completable {
        items.mapIndexed { index, card -> card.position = index }
        return cardRepository.save(items)
                .subscribeOn(Schedulers.io())
    }

    fun removeCard(card: Card): Completable = cardRepository.remove(card)
            .andThen(holderRepository.updateCounts(card.holder))
            .subscribeOn(Schedulers.io())

    private fun transform(card: Card): Card = card.apply {
        generatedBackgroundSeed = System.nanoTime()
    }

    private fun createCardExistError(): Throwable =
            ExceptionFactory.createError(ExceptionFactory.ErrorType.CARD_EXIST,
                    R.string.error_card_exist)
}
